[sw, dif] introduce reset manager DIF library

silvestrst opened the pull request:
This is the initial incomplete Reset Manager DIF implementation, also please see #3089 .

Design decisions:
* Get complete reset causes information, instead of single bit access routine (the causes are not mutually exclusive, and it is likely that the caller needs to know more than one/all of the causes to make an informative decision on what to do).
* Clear the whole reset causes field, instead of single bit clear routines.

Note: the peripheral does not have any information in the hardware dashboard regarding the V, D stage of the IP, and there is no documentation yet. This makes it harder to understand the use-cases.

silvestrst commented:
Forgot to run through `clang-format`, however it doesn't matter much as at this point I would like the header design evaluation rather than a deep review.

gkelly commented on sw/device/lib/dif/dif_resetmgr.h:
```diff
@@ -0,0 +1,187 @@
+ * Reset Manager reset causes information.
+ */
+typedef struct dif_resetmgr_causes {
+  bool cause[kDifResetmgrCauseLast + 1];
```
I'm not sold on the ergonomics of `bool[N]` to represent N different reset causes. If the total number of resets will always be strictly less than 32 then let's encode these as one-hot enumerations. I know the difference is between:

```c
if (cause.cause[kDifResetmgrCausePowerUp]) {
```

And

```c
if (cause & kDifResetmgrCausePowerUp) {
```

on a normal single-condition check. But since the bits aren't mutually exclusive I anticipate that the access pattern will be "did we reset because of X and only X?" and that two `ands` and an `or` if it's packed into a u32. If it's an array we have to iterate and accumulate the conditions.

Does that seem reasonable?

silvestrst replied:
> If the total number of resets will always be strictly less than 32
That was one of the reasons I choose the array of bools, but I think you are right - I can't imagine there ever would even be 32 reset causes, nevermind more.

Indeed, this is something I also considered. I thought that array of bools was more suggestive to how it is supposed to be used. However, like you said - `cause.cause[kDifResetmgrCausePowerUp]` and `cause & kDifResetmgrCausePowerUp` are not too much different, and `cause & kDifResetmgrCausePowerUp` is more ergonomic of the two.

silvestrst replied:
> If it's an array we have to iterate and accumulate the conditions.

Maybe I am misunderstanding, but we could simply do:
```
if ((causes.cause[kDifResetmgrCausePowerUp] && causes.cause[kDifResetmgr*]) ||
      causes.cause[kDifResetmgr*]) {
```

gkelly replied:
The condition I'm suggesting checking for is "our cause was _only_ PowerUp", for example...

```c
// Array of bools. Testing for _only_ kDifResetmgrCausePowerup.
if (!causes.cause[kDifResetmgrCausePowerUp]) {
  return false;
}
for (int i = 0; i < kDifResetmgrCauseCount; ++i) {
  if (i == kDifResetmgrCausePowerUp) {
    continue;
  }
  if (causes.cause[i] && i != (dif_cause_t)kDifResetmgrCausePowerUp) {
    return false;
  }
}
return true;

// Bitfields. Testing for _only_ kDifResetmgrCausePowerup.
return cause == kDifResetmgrCausePowerUp;
```

silvestrst replied:
I see, yeah, that makes sense.

gkelly commented on sw/device/lib/dif/dif_resetmgr.h:
```diff
@@ -0,0 +1,187 @@
+  dif_resetmgr_t *resetmgr, dif_resetmgr_peripheral_t peripheral,
+  dif_resetmgr_enable_t enable);
+
+typedef enum dif_resetmgr_peripheral_reset_result {
```
Missing doxygen comment for this type.

silvestrst replied:
Done.

gkelly commented on sw/device/lib/dif/dif_resetmgr.c:
```diff
@@ -0,0 +1,112 @@
+ * Note: reset causes are not mutually exclusive, so are implemented as a
+ *       bitfield.
+ */
+static const uint32_t resetmgr_causes_map[] = {
+  [kDifResetmgrCausePowerUp] = RSTMGR_RESET_INFO_CAUSE_POR,
+  [kDifResetmgrCauseLowPowerExit] = RSTMGR_RESET_INFO_CAUSE_LOW_POWER_EXIT,
+  [kDifResetmgrCauseWatchdog] = RSTMGR_RESET_INFO_CAUSE_WATCHDOG,
+  [kDifResetmgrCauseSecurityEscalation] = RSTMGR_RESET_INFO_CAUSE_ESCALATION,
+  [kDifResetmgrCauseNonDebugModuleRequest] =
+  RSTMGR_RESET_INFO_CAUSE_NON_DEBUG_MODULE,
+};
```
This and the `dif_resetmgr_causes_get` function get much simpler if you move this to a packed bitfield. Having now read the hjson for the hardware this always fits in five bits, which inclines me towards bitfields even more here.

silvestrst replied:
Will change to bitfields in the next push, as per previous comment.

silvestrst replied:
> This and the dif_resetmgr_causes_get function get much simpler if you move this to a packed bitfield.

Not sure, as we still have to map public enum to private register field offsets. We cannot use the register definitions in the `.h` file. We could hardcode the bit indexes, as:

```
typedef enum dif_resetmgr_cause {
  kDifResetmgrCausePowerUp = 1,
  kDifResetmgrCauseLowPowerExit = 2,
  kDifResetmgrCauseWatchdog = 4,
  kDifResetmgrCauseSecurityEscalation = 8,
  kDifResetmgrCauseNonDebugModuleRequest = 16,
  kDifResetmgrCauseLast = kDifResetmgrCauseNonDebugModuleRequest,
} dif_resetmgr_cause_t;
```
to be the same as:
```
#define RSTMGR_RESET_INFO_CAUSE_POR 0x1
#define RSTMGR_RESET_INFO_CAUSE_LOW_POWER_EXIT 0x2
#define RSTMGR_RESET_INFO_CAUSE_WATCHDOG 0x4
#define RSTMGR_RESET_INFO_CAUSE_ESCALATION 0x8
#define RSTMGR_RESET_INFO_CAUSE_NON_DEBUG_MODULE 0x10
```

But to me that doesn't seem all that great.

gkelly replied:
I think there's no requirement to have the values be the same. If we have a function that does:

```c
if (hw_cause & RSTMGR_RESET_INFO_CAUSE_POR) {
  cause |= kDifResetmgrCausePowerUp;
}
```

Then I think that's a perfectly reasonable conversion.

silvestrst replied:
Okay, sounds good.

silvestrst replied:
@lenary just suggested to make the values match the autogenerated definitions in the source file with static_assert. I think that is probably a better solution, so I will do that (avoids unnecessary runtime matching).

silvestrst commented:
@gkelly , what do you think of adding something like the following:
```
/**                                                                                                                                                                                                                                                            
 * Reset Manager configuration data.                                                                                                                                                                                                                           
 */
typedef struct dif_rstmgr_config {
  bool peripheral_reset_enable[kDifRstmgrPeripheralLast + 1];
} dif_rstmgr_config_t;

// TODO - description
DIF_WARN_UNUSED_RESULT
dif_rstmgr_result_t dif_rstmgr_configure(const dif_rstmgr_t *rstmgr,
                                                                 const dif_rstmgr_config_t *config,
                                                                 bool reset);
```

And removing `dif_rstmgr_peripheral_reset_enable` - I can imagine that the user would want to configure the peripheral enable once, and won't want to toggle it after. I am happy either way.

silvestrst commented:
@gkelly I have addressed your comments. To make the `bitfields` cleaner I could hard-code them the same as the register definitions and then verify that they match in the `*_unittest.cc`. Please let me know which of the different approaches you are most in favour and ill implement that. So far I think my choice would be `bool` array, but I am happy to go with any approach.

jon-flatley commented on sw/device/lib/dif/dif_rstmgr.c:
```diff
@@ -0,0 +1,108 @@
+
+// Register fields that are used in multiple functions.
+
+static const bitfield_field32_t rstmgr_causes_field = {
```
Use `kConst` naming scheme.

silvestrst replied:
Thanks.

silvestrst replied:
Done.

jon-flatley commented on sw/device/lib/dif/dif_rstmgr.h:
```diff
@@ -0,0 +1,185 @@
+ * TODO: no documentation currently available, would be useful
+ *       for writing a more detailed description for the individual causes.
+ */
+typedef enum dif_rstmgr_cause {
```
I'm not a fan of using an enum here. What do you think about
```
typedef uint32_t dif_rstmgr_cause_t;
const dif_rstmgr_cause_t kDifRstmgrCausePowerUp = 1;
const dif_rstmgr_cause_t kDifRstmgrCauseLowPowerExit = 2;
...
```

If you want  to stick with enums how about using a utility function to resolve the register offsets rather than building them into the enum values?

silvestrst replied:
I don't mind, can go either way, could you however explain why you prefer `const dif_rstmgr_cause_t kDifRstmgrCausePowerUp = 1`? Do you see a potential problem with encoding the values in an enum?

> utility function to resolve the register offsets rather than building them into the enum values?

The enum bit field values in this case only happen to match the offsets in the register definition. In reality there are no guarantees that they will match, hence we do the mapping in the `*_get` function (hopefully I understood your comment correctly).

gkelly commented on sw/device/lib/dif/dif_rstmgr.h:
```diff
@@ -0,0 +1,185 @@
+  kDifRstmgrPeripheralLast = kDifRstmgrPeripheralSpi,
+} dif_rstmgr_peripheral_t;
+
+typedef uint32_t dif_rstmgr_causes_bitfield_t;
```
Add doxygen.

silvestrst replied:
Done.

silvestrst commented:
Thank you @gkelly and @jon-flatley , I will aim to address your review comments tomorrow and will push this change for a wider review.

silvestrst commented:
TODO API:
* Add regen status get function.

TODO implementation:
* Use bit fields library in appropriate places.
* Avoid hardcoded `1`s and `0`s, and instead give them a meaningful `#define name`

silvestrst commented on sw/device/lib/dif/dif_rstmgr.c:
```diff
@@ -0,0 +1,161 @@
+};
+
+// Regen registers are "write 1 to clear".
+static const ptrdiff_t kRstmgrRegenOffsetsMap[kDifRstmgrPeripheralLast + 1] = {
```
Note: this and the next `*Map` arrays might become obsolete if `regen` and `reset` control bits are packed in a single `regen` and `reset` registers, instead of the current one register per peripheral. Please see #3089 for further information.

silvestrst replied:
Marking this as solved.

lenary commented on sw/device/lib/dif/dif_rstmgr.h:
```diff
@@ -0,0 +1,206 @@
+/**
+ * Reset Manager configuration data.
+ */
+typedef struct dif_rstmgr_config {
+  /**
+   * Reset Manager peripheral software reset lock configuration (`true` to lock,
+   * otherwise nothing, as the enable registers cannot be unlocked).
+   */
+  bool peripheral_software_reset_lock[kDifRstmgrPeripheralLast + 1];
+} dif_rstmgr_config_t;
```
I don't think of the reset manager as something you lock at configure-time. Instead provide two methods: one to lock a peripheral, and one to check if a peripheral is locked (i.e. with a bool out-param).

silvestrst replied:
Done.

lenary commented on sw/device/lib/dif/dif_rstmgr.h:
```diff
@@ -0,0 +1,206 @@
+ * @return `dif_rstmgr_result_t`.
+ */
+DIF_WARN_UNUSED_RESULT
+dif_rstmgr_software_reset_result_t dif_rstmgr_software_reset(
+    const dif_rstmgr_t *rstmgr, dif_rstmgr_peripheral_t peripheral,
+    dif_rstmgr_software_reset_t reset);
```
Is there a way of reading the current `hold in reset` value, or is the register write-only?

silvestrst replied:
There is, I will add a routine to check it.

silvestrst commented:
* I think I have addressed all of the outstanding review comments.
* It is a bit awkwards with separate regen and reset registers for every peripheral, please see #3089 for context.

silvestrst commented on sw/device/lib/dif/dif_rstmgr.h:
```diff
@@ -0,0 +1,239 @@
+ *
+ * This type should be treated as opaque by users.
+ */
+typedef struct dif_rstmgr { dif_rstmgr_param_t params; } dif_rstmgr_t;
```
Hey @alphan thank you for the review! I will fix up the comments next time I am working on this DIF.

Regarding this specific comment, I just copy/pasted this from the DIF template by @mcy , which has landed recently.

silvestrst replied:
> Regarding this specific comment, I just copy/pasted this from the DIF template by @mcy , which has landed recently.

I think this is also true for some other comments.

mcy replied:
Should be `params` not `param` (typo in the template... :X). But this is otherwise correct.

silvestrst replied:
Done.

silvestrst commented on sw/device/lib/dif/dif_rstmgr.h:
```diff
@@ -0,0 +1,239 @@
+  kDifRstmgrCauseWatchdog = 0x4,
+  kDifRstmgrCauseSecurityEscalation = 0x8,
+  kDifRstmgrCauseNonDebugModuleRequest = 0x10,
+} dif_rstmgr_cause_t;
```
That it indeed a good idea, thanks!

silvestrst replied:
Done.

mcy replied:
I would actually remove the parentheses, since they're superfluous.

silvestrst commented on sw/device/lib/dif/dif_rstmgr.c:
```diff
@@ -0,0 +1,215 @@
+                   RSTMGR_RST_USB_N_RST_USB_N,
+               "SPI and USB reset bit offsets are not the same");
+
+#define RSTMGR_PERIPHERAL_REGEN RSTMGR_SPI_DEVICE_REGEN_SPI_DEVICE_REGEN
+#define RSTMGR_PERIPHERAL_RST_N RSTMGR_RST_SPI_DEVICE_N_RST_SPI_DEVICE_N
```
Sounds good.

silvestrst replied:
Done.

silvestrst commented on sw/device/lib/dif/dif_rstmgr.c:
```diff
@@ -0,0 +1,215 @@
+    return kDifRstmgrBadArg;
+  }
+
+  mmio_region_t base_addr = handle->params.base_addr;
```
Yup, I understand, I just thought that because @mcy  change (DIF template) has landed - everyone has agreed on it, so I simply copy/pasted it (I admit, I haven't followed the that PR closely).

silvestrst replied:
I think we are going forward with Miguel's template, so I will mark this comment as addressed (keeping the params).

silvestrst commented on sw/device/lib/dif/dif_rstmgr.c:
```diff
@@ -0,0 +1,215 @@
+
+  // De-assert software resets.
+  for (dif_rstmgr_peripheral_t i = 0; i <= kDifRstmgrPeripheralLast; ++i) {
+    mmio_region_write32(base_addr, kRstmgrResetOffsetsMap[i], 1);
```
Indeed, thanks!

silvestrst replied:
Done.

silvestrst commented on sw/device/lib/dif/dif_rstmgr.c:
```diff
@@ -0,0 +1,215 @@
+  mmio_region_write32(base_addr, reg_offset, bitfield);
+}
+
+static void rstmgr_software_reset_lock(mmio_region_t base_addr,
+                                       ptrdiff_t reg_offset) {
+  uint32_t bitfield = bitfield_bit32_write(0, RSTMGR_PERIPHERAL_REGEN, true);
+  mmio_region_write32(base_addr, reg_offset, bitfield);
+}
```
Might still be worth keeping in a separate function for simetricity with `*_release`.

alphan replied:
Your call, just wanted to point out.

silvestrst commented on sw/device/lib/dif/dif_rstmgr.c:
```diff
@@ -0,0 +1,215 @@
+  mmio_region_write32(base_addr, reg_offset, bitfield);
+}
+
+static bool rstmgr_software_reset_is_held(mmio_region_t base_addr,
+                                          ptrdiff_t reg_offset) {
+  uint32_t bitfield = mmio_region_read32(base_addr, reg_offset);
+  return !bitfield_bit32_read(bitfield, RSTMGR_PERIPHERAL_RST_N);
+}
```
Same as above, however I don't really have a strong opinion on this.

alphan replied:
Your call, just wanted to point out.

silvestrst commented on sw/device/lib/dif/dif_rstmgr.c:
```diff
@@ -0,0 +1,215 @@
+        [kDifRstmgrPeripheralUsb] = RSTMGR_RST_USB_N_REG_OFFSET,
+};
+
+static void rstmgr_software_reset_hold(mmio_region_t base_addr,
```
I will look into it.

silvestrst replied:
Very good suggestion, thanks!

Done.

silvestrst commented on sw/device/lib/dif/dif_rstmgr.c:
```diff
@@ -0,0 +1,215 @@
+  ptrdiff_t reset_offset = kRstmgrResetOffsetsMap[peripheral];
+  switch (reset) {
+    case kDifRstmgrSoftwareReset:
+      rstmgr_software_reset_hold(base_addr, reset_offset);
+      rstmgr_software_reset_release(base_addr, reset_offset);
```
Good point, hard to say - I don't think there is documentation yet for this peripheral, and I didn't dig into the RTL. Valid point though. Maybe @msfschaffner or @tjaychen could clarify this for us (thanks)?

silvestrst replied:
> I don't think there is documentation 

My bad, found it under `hw`

silvestrst replied:
Resolving this, will chase it up separately.

silvestrst commented on sw/device/lib/dif/dif_rstmgr.c:
```diff
@@ -0,0 +1,215 @@
+  return kDifRstmgrOk;
+}
+
+dif_rstmgr_result_t dif_rstmgr_software_reset_asserted(
```
Yeah, I wasn't sure about the naming here either. Maybe `...is_held`?

mcy replied:
I like `is_held`!

silvestrst replied:
Done, `*_is_held` it is :)

silvestrst commented on sw/device/lib/dif/dif_rstmgr.h:
```diff
@@ -0,0 +1,239 @@
+ * @param asserted 'true' when held in reset, `false` otherwise.
+ * @return The result of the operation.
+ */
+DIF_WARN_UNUSED_RESULT
+dif_rstmgr_result_t dif_rstmgr_software_reset_asserted(
+    const dif_rstmgr_t *handle, dif_rstmgr_peripheral_t peripheral,
+    bool *asserted);
```
Yeah, I see your point, I guess because there is an extra enum variant it is better to just keep it as bool?

alphan replied:
Yes, I agree.

silvestrst commented on sw/device/lib/dif/dif_rstmgr.h:
```diff
@@ -0,0 +1,239 @@
+                                    dif_rstmgr_t *handle);
+
+/**
+ * Resets the Reset Manager registers to sane defaults.
```
Yeah, I was struggling to describe it. Also, I think this function needs a check for whether these are locked, otherwise they won't be able to be de-asserted. Maybe this function should be completely renamed.

silvestrst commented on sw/device/lib/dif/dif_rstmgr.h:
```diff
@@ -0,0 +1,239 @@
+
+/**
+ * @file
+ * @brief <a href="">Reset Manager</a> Device Interface Functions
```
Yes you are right, didn't realise there was doc under 'hw', oops!

silvestrst replied:
There is no documentation for this peripheral yet.

lenary replied:
Just link to where it will be, which will be at `/hw/ip/rstmgr/doc/`.

silvestrst replied:
Done.

silvestrst commented on sw/device/lib/dif/dif_rstmgr.h:
```diff
@@ -0,0 +1,239 @@
+   * Indicates that some parameter passed into a function failed a
+   * precondition.
+   *
+   * When this value is returned, no hardware operations occured.
```
Thanks, these would need fixing in the DIF template as well.

silvestrst commented on sw/device/lib/dif/dif_rstmgr.h:
```diff
@@ -0,0 +1,239 @@
+/**
+ * A toggle state: enabled, or disabled.
+ *
+ * This enum may be used instead of a `bool` when describing an
+ * enabled/disabled state.
```
You are right, this is not used anywhere.

Removed.

silvestrst commented on sw/device/lib/dif/dif_rstmgr.h:
```diff
@@ -0,0 +1,239 @@
+extern "C" {
+#endif  // __cplusplus
+
+/**
+ * A toggle state: enabled, or disabled.
+ *
+ * This enum may be used instead of a `bool` when describing an
+ * enabled/disabled state.
+ */
+typedef enum dif_rstmgr_toggle {
+  /**
+   * The "enabled" state.
+   */
+  kDifRstmgrToggleEnabled,
+  /**
+   * The "disabled" state.
+   */
+  kDifRstmgrToggleDisabled,
+} dif_Rstmgr_toggle_t;
```
Done, as per above comment.

silvestrst commented on sw/device/lib/dif/dif_rstmgr.h:
```diff
@@ -0,0 +1,239 @@
+typedef uint32_t dif_rstmgr_causes_bitfield_t;
+
+/**
+ * Reset Manager all possible reset causes enumeration.
```
Would you like me to use something else instead of "Reset Manager", could you please expand a little on your comment?

alphan replied:
I thought it doesn't read like a proper sentence, e.g. "All possible reset causes".

alphan replied:
Please feel free to ignore.

silvestrst replied:
This has been renamed in the follow-up SW HW sync change. This is not applicable anymore - resolving!

silvestrst commented on sw/device/lib/dif/dif_rstmgr.h:
```diff
@@ -0,0 +1,239 @@
+/**
+ * Locks out requested peripheral reset functionality.
+ *
+ * This function is reentrant: calling it while functionality is locked will
```
Agree, done.

Probably will need to be changed in the DIF template as well.

silvestrst replied:
This has been done before the change was merged. Marking as resolved.

silvestrst commented on sw/device/lib/dif/dif_rstmgr.h:
```diff
@@ -0,0 +1,239 @@
+DIF_WARN_UNUSED_RESULT
+dif_rstmgr_result_t dif_rstmgr_reset_lock(const dif_rstmgr_t *handle,
+                                          dif_rstmgr_peripheral_t peripheral);
+/**
```
Done.

silvestrst commented:
Fixed up most of @alphan suggestions, some left open for further discussion.

lenary commented:
This PR should not block on #3089 - we can always land implementation changes later.

Again, look to see what review comments can be pushed into a follow-up PR so this can be landed sooner.

mcy commented on sw/device/lib/dif/dif_rstmgr.h:
```diff
@@ -0,0 +1,223 @@
+ * Reset Manager software reset available peripherals.
+ */
+typedef enum dif_rstmgr_peripheral {
+  kDifRstmgrPeripheralSpi = 0,
+  kDifRstmgrPeripheralUsb,
+  kDifRstmgrPeripheralLast = kDifRstmgrPeripheralUsb,
```
These are missing comments (and the last one should be marked \internal).

silvestrst replied:
Done.

mcy commented on sw/device/lib/dif/dif_rstmgr.c:
```diff
@@ -0,0 +1,215 @@
+};
+
+// Reset registers, when set to 0, peripheral is held in reset.
+static const ptrdiff_t kRstmgrResetOffsetsMap[kDifRstmgrPeripheralLast + 1] = {
```
AIUI, you don't need to include the array extent on a global; C will just infer it. (I used to believe this was C++-only but it seems to work correctly in C, too.)

silvestrst replied:
Done.

silvestrst commented:
# I propose to merge this PR with all the outstanding comments and other actions resolved in the follow-up. Please see #3454 .

@gkelly @lenary @alphan , please "approve" this change if you are happy with the proposed strategy.

gkelly commented:
> # I propose to merge this PR with all the outstanding comments and other actions resolved in the follow-up. Please see #3454 .
> @gkelly @lenary @alphan , please "approve" this change if you are happy with the proposed strategy.

I'm happy with that approach, approving.

gkelly commented:
With @lenary on vacation I've looked at his requested changes and am satisfied that @silvestrst has addressed them or referenced them in the followup issue.
